src/Sentry/Internal/CachingBackgroundWorker.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Sentry.Extensibility;
+using Sentry.Protocol.Envelopes;
+
+namespace Sentry.Internal
+{
+    internal class CachingBackgroundWorker : BackgroundWorkerBase
+    {
+        private readonly DirectoryInfo _cacheDirectory;
+
+        public CachingBackgroundWorker(ITransport transport, SentryOptions options)
+            : base(transport, options)
+        {
+            _cacheDirectory =
+                options.CacheDirectory ??
+                throw new InvalidOperationException("Cache directory is not set.");
+        }
+
+        private IEnumerable<FileInfo> GetCachedEnvelopeFiles()
+        {
+            if (!_cacheDirectory.Exists)
+            {
+                return Enumerable.Empty<FileInfo>();
+            }
+
+            return _cacheDirectory.EnumerateFiles("*.envelope");
+        }
+
+        private FileInfo? TryGetNextEnvelopeFile() => GetCachedEnvelopeFiles()
+            .OrderBy(f => f.CreationTimeUtc)
+            .FirstOrDefault();
+
+        private async ValueTask FlushCacheAsync(CancellationToken cancellationToken = default)
+        {
+            while (TryGetNextEnvelopeFile() is { } envelopeFile)
+            {
+                Options.DiagnosticLogger?.LogDebug("Sending cached envelope: {0}", envelopeFile.FullName);
+
+                using var envelope = await Envelope.DeserializeAsync(
+                    envelopeFile.OpenRead(),
+                    cancellationToken
+                ).ConfigureAwait(false);
+
+                await Transport.SendEnvelopeAsync(envelope, cancellationToken).ConfigureAwait(false);
+
+                envelopeFile.Delete();
+            }
+        }
+
+        private async ValueTask CacheEnvelopeAsync(Envelope envelope, CancellationToken cancellationToken = default)
+        {
+            if (GetCachedEnvelopeFiles().Count() >= 50)
+            {
+                return;
+            }
+
+            var envelopeFile = new FileInfo(
+                Path.Combine(
+                    _cacheDirectory.FullName,
+                    $"{DateTimeOffset.Now}_{envelope.TryGetEventId()}_{envelope.GetHashCode()}")
```

Comment:
You used a `.envelope` extension above, we need it here?

What about UtcNow so the file names are not subjective to daylight savings/timezone changes?

What if not event id in the envelope, maybe `?? ...`

Reply:
Good points. I didn't think this through very much yet.

Reply:
Fixed.

If event ID is not set then it will be empty, so it will look like this:

```
123456789__91231293.envelope
```

Note the double underscore. Since the file names are not really meant to be read by users, I think it's okay. It will still be sorted by date and combination of unix time + hash code will ensure there are no collisions. Event ID is there for debugging purposes, mostly.